doc/man/stdio.3
.\"
.\" Copyright (C) 2000  Internet Software Consortium.
.\"
.\" Permission to use, copy, modify, and distribute this document for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND INTERNET SOFTWARE CONSORTIUM
.\" DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL
.\" INTERNET SOFTWARE CONSORTIUM BE LIABLE FOR ANY SPECIAL, DIRECT,
.\" INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING
.\" FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
.\" NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
.\" WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\" $Id: stdio.3,v 1.1 2000/06/23 00:30:12 jim Exp $
.\"
.Dd Jun 30, 2000
.Dt STDIO 3
.Os BIND9 9
.ds vT BIND9 Programmer's Manual
.Sh NAME
.Nm isc_stdio_open ,
.Nm isc_stdio_close ,
.Nm isc_stdio_seek ,
.Nm isc_stdio_read ,
.Nm isc_stdio_write ,
.Nm isc_stdio_flush ,
.Nm isc_stdio_sync
.Nd BIND9 interface to stdio functions
.Sh SYNOPSIS
.Fd #include <config.h>

.Fd #include <errno.h>
.Fd #include <unistd.h>

.Fd #include <isc/stdio.h>

.Fd #include \*qerrno2result.h\*q

.Ft isc_result_t
.Fn isc_stdio_open "const char *filename" "const char *mode"  "FILE **fp"
.Ft isc_result_t
.Fn isc_stdio_close "FILE *f"
.Ft isc_result_t
.Fn isc_stdio_seek "FILE *f" "long offset" "int whence"
.Ft isc_result_t
.Fo isc_stdio_read
.Fa "void *ptr"
.Fa "size_t size"
.Fa "size_t nmemb"
.Fa "FILE *f"
.Fa "size_t *nret"
.Fc
.Ft isc_result_t
.Fo isc_stdio_write
.Fa "void *ptr"
.Fa "size_t size"
.Fa "size_t nmemb"
.Fa "FILE *f"
.Fa "size_t *nret"
.Fc
.Ft isc_result_t
.Fn isc_stdio_flush "FILE *f"
.Ft isc_result_t
.Fn isc_stdio_sync "FILE *f"
.Sh DESCRIPTION
The BIND9 library uses the following routines for standard I/O
operations.
These functions call their equivalents in the operating system's stdio
library.
.Pp
.Fn isc_stdio_open
opens the file
.Fa filename
with access mode
.Fa mode .
The string
.Fa mode
should correspond to the
.Fa mode
parameter in
.Xr fopen 3 .
If the file open succeeds,
.Fa fp
contains a pointer to the pointer to the relevant stdio
.Dv FILE
structure.
Files get closed by calling
.Fn isc_stdio_close .
.Fa fp
should be a pointer to the
.DV FILE
structure that was returned by an earlier call to
.Fn isc_stdio_open .
.Pp
The file position indicator for a stdio stream is adjusted with
.Fn isc_stdio_seek .
.Fa f
is the
.Dv FILE
pointer and
.Fa offset
and 
.Fa whence
have the usual meanings.
The new position is found by adding
.Fa offset
bytes to the position specified by
.Fa whence 
which is set to the values
.Dv SEEK_SET ,
.Dv SEEK_CUR ,
or
.Dv SEEK_END .
These values indicate the new position is calculated relative to the start
of the file, the current position indicator, or end of file
respectively.
.Pp
Reading and writing on a stdio stream is performed using
.Fn isc_stdio_read
and
.Fn isc_stdio_write
respectively.
The function
.Fn isc_stdio_read
reads reads
.Fa nmemb
objects, each
.Fa size
bytes long, from the stream pointed to by
.Fa f .
The objects are stored at the location given by
.Fa ptr .
.Fa isc_stdio_write
writes 
.Fa nmemb
objects, each
.Fa size
bytes long, from location
.Fa ptr
to the stream
.Fa f .
The number of objects written is returned via
.Fa nret .
If this is less than
.Fa nmemb ,
some type of write error occurred.
.Pp
A
.Xr write 2
of buffered data associated with the output
.Fa f
can be forced with
.Fn isc_stdio_flush .
This data can be committed to permanent storage by
.Fn isc_stdio_sync
which will invoke
.Xr fsync 2 .
.Sh RETURN VALUES
Successful calls to all these functions return
.Er ISC_R_SUCCESS .
If the calls fail, the error code returned by the operating system is
mapped to the appropriate BIND9 error code by calling
.Xr isc__errno2result 3 .
.Fn isc_stdio_read
returns
.Er ISC_R_EOF
at end of file.
.Sh SEE ALSO
.Xr stdio 3 ,
.Xr fopen 3 ,
.Xr fclose 3 ,
.Xr fread 3 ,
.Xr fwrite 3 ,
.Xr fflush 3 ,
.Xr fsync 2 ,
.Xr isc__errno2result 3 .


NEWS
doc/arm/changelog.rst

lib/isc/stdio.c
/*
 * Copyright (C) Internet Systems Consortium, Inc. ("ISC")
 *
 * SPDX-License-Identifier: MPL-2.0
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, you can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * See the COPYRIGHT file distributed with this work for additional
 * information regarding copyright ownership.
 */

#include <errno.h>
#include <unistd.h>

#include <isc/stat.h>
#include <isc/stdio.h>
#include <isc/util.h>

#include "errno2result.h"

isc_result_t
isc_stdio_open(const char *filename, const char *mode, FILE **fp) {
	FILE *f;

	f = fopen(filename, mode);
	if (f == NULL) {
		return isc__errno2result(errno);
	}
	*fp = f;
	return ISC_R_SUCCESS;
}

isc_result_t
isc_stdio_close(FILE *f) {
	int r;

	r = fclose(f);
	if (r == 0) {
		return ISC_R_SUCCESS;
	} else {
		return isc__errno2result(errno);
	}
}

isc_result_t
isc_stdio_seek(FILE *f, off_t offset, int whence) {
	int r;

	r = fseeko(f, offset, whence);
	if (r == 0) {
		return ISC_R_SUCCESS;
	} else {
		return isc__errno2result(errno);
	}
}

isc_result_t
isc_stdio_tell(FILE *f, off_t *offsetp) {
	off_t r;

	REQUIRE(offsetp != NULL);

	r = ftello(f);
	if (r >= 0) {
		*offsetp = r;
		return ISC_R_SUCCESS;
	} else {
		return isc__errno2result(errno);
	}
}

isc_result_t
isc_stdio_read(void *ptr, size_t size, size_t nmemb, FILE *f, size_t *nret) {
	isc_result_t result = ISC_R_SUCCESS;
	size_t r;

	clearerr(f);
	r = fread(ptr, size, nmemb, f);
	if (r != nmemb) {
		if (feof(f)) {
			result = ISC_R_EOF;
		} else {
			result = isc__errno2result(errno);
		}
	}
	if (nret != NULL) {
		*nret = r;
	}
	return result;
}

isc_result_t
isc_stdio_write(const void *ptr, size_t size, size_t nmemb, FILE *f,
		size_t *nret) {
	isc_result_t result = ISC_R_SUCCESS;
	size_t r;

	clearerr(f);
	r = fwrite(ptr, size, nmemb, f);
	if (r != nmemb) {
		result = isc__errno2result(errno);
	}
	if (nret != NULL) {
		*nret = r;
	}
	return result;
}

isc_result_t
isc_stdio_flush(FILE *f) {
	int r;

	r = fflush(f);
	if (r == 0) {
		return ISC_R_SUCCESS;
	} else {
		return isc__errno2result(errno);
	}
}

/*
 * OpenBSD has deprecated ENOTSUP in favor of EOPNOTSUPP.
 */
#if defined(EOPNOTSUPP) && !defined(ENOTSUP)
#define ENOTSUP EOPNOTSUPP
#endif /* if defined(EOPNOTSUPP) && !defined(ENOTSUP) */

isc_result_t
isc_stdio_sync(FILE *f) {
	struct stat buf;
	int r;

	if (fstat(fileno(f), &buf) != 0) {
		return isc__errno2result(errno);
	}

	/*
	 * Only call fsync() on regular files.
	 */
	if ((buf.st_mode & S_IFMT) != S_IFREG) {
		return ISC_R_SUCCESS;
	}

	r = fsync(fileno(f));
	if (r == 0) {
		return ISC_R_SUCCESS;
	} else {
		return isc__errno2result(errno);
	}
}
